kernel/thread.c

```diff
@@ -581,29 +801,36 @@ static long _run_thread(void* arg_)
                 myst_release_process_mappings(thread->pid);
             }
 
-            free(thread->main.cwd);
-            thread->main.cwd = NULL;
-
-            /* Passed in by SYS_myst_unmap_on_exit */
-            if (thread->unmap_on_exit_addr)
             {
-                myst_munmap(
-                    thread->unmap_on_exit_addr, thread->unmap_on_exit_length);
+                size_t i = thread->main.unmap_on_exit_used;
+                while (i)
+                {
+                    myst_munmap(
+                        thread->main.unmap_on_exit[i - 1].ptr,
+                        thread->main.unmap_on_exit[i - 1].size);
+                    i--;
+                }
             }
 
+            free(thread->main.cwd);
+            thread->main.cwd = NULL;
+
             procfs_pid_cleanup(thread->pid);
         }
 
-        myst_zombify_thread(thread);
-
         /* Send a SIGCHLD to the parent process */
-        myst_syscall_kill(thread->ppid, SIGCHLD);
+        if (!is_child_thread)
+        {
+            myst_syscall_kill(thread->ppid, SIGCHLD);
```

Comment:
If this is scoped to a process thread, it should move into the previous `if (!is_child_thread)` branch, i.e., after line 818.

Reply:
done